extensions/controllers/sandboxwarmpool_controller.go

```diff
@@ -0,0 +1,260 @@
+)
+
+const (
+	poolLabel = "agents.x-k8s.io/pool"
+)
+
+// SandboxWarmPoolReconciler reconciles a SandboxWarmPool object
+type SandboxWarmPoolReconciler struct {
+	client.Client
+}
+
+//+kubebuilder:rbac:groups=extensions.agents.x-k8s.io,resources=sandboxwarmpools,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=extensions.agents.x-k8s.io,resources=sandboxwarmpools/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile implements the reconciliation loop for SandboxWarmPool
+func (r *SandboxWarmPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// Fetch the SandboxWarmPool instance
+	warmPool := &extensionsv1alpha1.SandboxWarmPool{}
+	if err := r.Get(ctx, req.NamespacedName, warmPool); err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Info("SandboxWarmPool resource not found. Ignoring since object must be deleted")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Failed to get SandboxWarmPool")
+		return ctrl.Result{}, err
+	}
+
+	// Handle deletion
+	if !warmPool.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("SandboxWarmPool is being deleted")
+		return ctrl.Result{}, nil
+	}
+
+	// Save old status for comparison
+	oldStatus := warmPool.Status.DeepCopy()
+
+	// Reconcile the pool (create or delete Pods as needed)
+	if err := r.reconcilePool(ctx, warmPool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Update status if it has changed
+	if err := r.updateStatus(ctx, oldStatus, warmPool); err != nil {
+		log.Error(err, "Failed to update SandboxWarmPool status")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcilePool ensures the correct number of pods exist in the pool
+func (r *SandboxWarmPoolReconciler) reconcilePool(ctx context.Context, warmPool *extensionsv1alpha1.SandboxWarmPool) error {
+	log := log.FromContext(ctx)
+
+	// TODO: Use a hash value for the poolLabelValue
+	poolLabelValue := warmPool.Name
+
+	// List all pods with the pool label
+	podList := &corev1.PodList{}
+	labelSelector := labels.SelectorFromSet(labels.Set{
+		poolLabel: poolLabelValue,
+	})
+
+	if err := r.List(ctx, podList, &client.ListOptions{
+		LabelSelector: labelSelector,
+		Namespace:     warmPool.Namespace,
+	}); err != nil {
+		log.Error(err, "Failed to list pods")
+		return err
+	}
+
+	// Filter pods by ownership and adopt orphans
+	var activePods []corev1.Pod
+	var allErrors error
+
+	for _, pod := range podList.Items {
+		// Skip pods that are being deleted
+		if !pod.ObjectMeta.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		// Get the controller owner reference
+		controllerRef := metav1.GetControllerOf(&pod)
+
+		if controllerRef == nil {
+			// Pod has no controller - adopt it
+			log.Info("Adopting orphaned pod", "pod", pod.Name)
+			if err := r.adoptPod(ctx, warmPool, &pod); err != nil {
```

Comment:
This is a corner case, if you just adopt a pod with label selected, but the template maybe totally different.